### ISSUE TITLE
Data race in std::unique_lock

### ISSUE THREAD

**Issue by TmLev** [llvm/llvm-project#49709]
|  |  |
| --- | --- |
| Bugzilla Link | [50365](https://llvm.org/bz50365) |
| Version | unspecified |
| OS | All |
| CC | @dwblaikie,@mclow |

## Extended Description 
There are two lines in ```unique_lock::unlock``` that should appear in reverse order:
https://github.com/llvm/llvm-project/blob/d480f968ad8b56d3ee4a6b6df5532d485b0ad01e/libcxx/include/__mutex_base#L255
https://github.com/llvm/llvm-project/blob/d480f968ad8b56d3ee4a6b6df5532d485b0ad01e/libcxx/include/__mutex_base#L256

MRE would require a thread pool with fibers running over it (that's how TSan detected the race condition in the first place) so that one thread has unlocked ```__m_``` but not written false yet, and another thread is destroying the lock already.

**Comment by Quuxplusone** [comment]
Are you saying the following...?

- You've got two threads A and B, both able to access the same shared `unique_lock` object `u`, and `u` wraps a mutex `m`.
- Thread A is going to run `u.lock(); u.unlock();`
- Simultaneously, Thread B is going to run `m.lock(); u.~unique_lock();`
- If A wins the initial contention, then B will block inside `m.lock()`. B will unblock as soon as `m` becomes available, while A is still in the middle of running `u.unlock()` (that is, `u` is still in use by Thread A). This means that A and B have a data race because they're both trying to modify `u` at the same time (A is trying to finish up `u.unlock()`; B is trying to destroy `u`.)

But that's not `unique_lock`'s fault, is it? Isn't that the fault of `A` and `B` for both trying to touch object `u` without synchronization? `u` wraps `m`, but `u` is not protected _by_ `m`. I'd call this Not A (libc++) Bug, but rather a data race (UB) in the user's own code.

Swapping the two lines in libc++ might suppress the symptoms in this particular user's code, but has a non-zero risk of causing other hard-to-reason-about effects on other (non-UB) code.

**Comment by quyse** [comment]
I believe I hit the same issue recently with LLVM 19/libc++, and it was frustrating to hunt down.

> But that's not `unique_lock`'s fault, is it? Isn't that the fault of `A` and `B` for both trying to touch object `u` without synchronization? `u` wraps `m`, but `u` is not protected _by_ `m`. I'd call this Not A (libc++) Bug, but rather a data race (UB) in the user's own code.

In my case `std::unique_lock` was a member of a work item class. Something like that:

```cpp
struct Work;

std::mutex mutex;
std::queue<Work*> workToDo;

struct Work
{
  Work()
  : lock_{mutex} {}

  void Enqueue()
  {
    workToDo.push(this);
    lock_.unlock();
  }

  std::unique_lock<std::mutex> lock_;
};

// thread 1

(new Work())->Enqueue();

// thread 2

std::unique_lock<mutex> lock{mutex};
Work* work = workToDo.front();
workToDo.pop();
DoSomething(work);
delete work;
```

I understand that `std::unique_lock` is not considered thread safe by itself, but I sort of expect `std::unique_lock::unlock` to be an atomic operation. That is, if `std::unique_lock` holds the mutex which also protects the lock itself, it should be fine, because only one thread at a time works with the lock: when the first thread calls `unlock`, it unlocks the mutex, and the second thread nicely takes over.

Now if `std::unique_lock::unlock` is not atomic, this of course falls apart. But atomicity of unlock looks like a reasonable assumption to me. Otherwise, what else I'm supposed to do here, add a dedicated mutex for unique_lock? 😄

**Comment by philnik777** [comment]
> I believe I hit the same issue recently with LLVM 19/libc++, and it was frustrating to hunt down.
> 
> > But that's not `unique_lock`'s fault, is it? Isn't that the fault of `A` and `B` for both trying to touch object `u` without synchronization? `u` wraps `m`, but `u` is not protected _by_ `m`. I'd call this Not A (libc++) Bug, but rather a data race (UB) in the user's own code.
> 
> In my case `std::unique_lock` was a member of a work item class. Something like that:
> 
> struct Work;
> 
> std::mutex mutex;
> std::queue<Work*> workToDo;
> 
> struct Work
> {
>   Work()
>   : lock_{mutex} {}
> 
>   void Enqueue()
>   {
>     workToDo.push(this);
>     lock_.unlock();
>   }
> 
>   std::unique_lock<std::mutex> lock_;
> };
> 
> // thread 1
> 
> (new Work())->Enqueue();
> 
> // thread 2
> 
> std::unique_lock<mutex> lock{mutex};
> Work* work = workToDo.front();
> workToDo.pop();
> DoSomething(work);
> delete work;
> 
> I understand that `std::unique_lock` is not considered thread safe by itself, but I sort of expect `std::unique_lock::unlock` to be an atomic operation. That is, if `std::unique_lock` holds the mutex which also protects the lock itself, it should be fine, because only one thread at a time works with the lock: when the first thread calls `unlock`, it unlocks the mutex, and the second thread nicely takes over.
> 
> Now if `std::unique_lock::unlock` is not atomic, this of course falls apart. But atomicity of unlock looks like a reasonable assumption to me. Otherwise, what else I'm supposed to do here, add a dedicated mutex for unique_lock? 😄

I'd argue that you should use `unique_lock` only ever locally. In this case, you'd have a `unique_lock` in `Enqueue` to lock `workToDo`. I'm not sure why you lock `workToDo` in the constructor or why you have the `unique_lock` as a member of your class.

**Comment by quyse** [comment]
> I'd argue that you should use `unique_lock` only ever locally. In this case, you'd have a `unique_lock` in `Enqueue` to lock `workToDo`. I'm not sure why you lock `workToDo` in the constructor or why you have the `unique_lock` as a member of your class.

Yeah, I looked at my code again and actually I agree, I can rewrite it to use only a local unique_lock variable. The original code ([here](https://github.com/quyse/coil-core/blob/b3899673b129a88843594fe257b8b02242ee51cb/src/coil/tasks_sync.cppm)) is an implementation of conditional variable for use with C++ coroutines. The intent was to lock a mutex in one function (`WaitAwaiter` constructor) and unlock it in another (`WaitAwaiter::await_suspend`), since I cannot pass additional arguments to `await_suspend`, I kept unique_lock as a member of the awaiter class.

So, yeah, in this particular case it seems I can simplify the code and move all locking into `await_suspend`, thanks for that! I guess my code example doesn't add much here then, sorry.

**Comment by ldionne** [comment]
Closing since this doesn't seem to be a libc++ bug per the above discussion.

**Comment by TmLev** [comment]
> Closing since this doesn't seem to be a libc++ bug per the above discussion.

But Marshall Clow actually kind of agreed it’s a bug:

> I concur; that's a bug.

Also, the opening post isn’t related to the discussion that’s developed here over the last few days.

**Comment by philnik777** [comment]
> > Closing since this doesn't seem to be a libc++ bug per the above discussion.
> 
> But Marshall Clow actually kind of agreed it’s a bug:
> 
> > I concur; that's a bug.
> 
> Also, the opening post isn’t related to the discussion that’s developed here over the last few days.

Since he didn't give any reasoning and is unlikely to give one (since he's not active in the community anymore) I don't think we can give that much weight. Given that there was no use-case in the original report (and not even a reproducer) I'm not sure we can say much beyond "please give us a reasonable use-case or some reasoning this is non-conforming".

**Comment by ldionne** [comment]
My understanding of the discussion in this issue is that we concluded the problem was in the user code accessing `unique_lock` in an unsynchronized way. I'm OK with reopening in case there is additional information, but at the moment I believe that closing this as "invalid" (problem in the user code) is the correct thing to do.
